protocol/directory.go

```diff
@@ -91,14 +109,77 @@ func (d *ConiksDirectory) Register(req *RegistrationRequest) (
 		// insert new data to the directory on-the-fly
 		tb, err := d.pad.TB(req.Username, req.Key)
 		if err != nil {
-			return nil, ErrorDirectory
+			return NewErrorResponse(ErrorDirectory), ErrorDirectory
 		}
 		d.tbs[req.Username] = tb
 		return NewRegistrationProof(ap, d.LatestSTR(), tb, Success)
 	} else {
 		if err = d.pad.Set(req.Username, req.Key); err != nil {
-			return nil, ErrorDirectory
+			return NewErrorResponse(ErrorDirectory), ErrorDirectory
 		}
 		return NewRegistrationProof(ap, d.LatestSTR(), nil, Success)
 	}
 }
+
+func (d *ConiksDirectory) KeyLookup(req *KeyLookupRequest) (
+	Response, ErrorCode) {
+	ap, err := d.pad.Lookup(req.Username)
+	if err != nil {
+		return NewErrorResponse(ErrorDirectory), ErrorDirectory
+	}
+
+	if !bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		// if not found in the tree, do lookup in tb array
+		if d.useTBs {
+			if tb := d.tbs[req.Username]; tb != nil {
+				return NewKeyLookupProof(ap, d.LatestSTR(), tb, Success)
+			}
+		}
+		return NewKeyLookupProof(ap, d.LatestSTR(), nil, ErrorNameNotFound)
+	}
+	return NewKeyLookupProof(ap, d.LatestSTR(), nil, Success)
+}
+
+func (d *ConiksDirectory) KeyLookupInEpoch(req *KeyLookupInEpochRequest) (
```

Comment:
> What request is the client making that requires that?
>> What if Bob goes offline for a couple epochs and needs to verify a message I sent him while he was offline?

> In other words, all the functionality of KeyLookupInEpoch is covered by Monitor now.

I think the main difference is that `KeyLookupInEpoch` would be used for looking up your contacts' key in past epochs, whereas `Monitor` would be used for monitoring your own binding. 

Maybe we want to discuss this in #70 ?